packages/lodestar/src/sync/initial/fast/index.ts

```diff
@@ -190,9 +189,10 @@ export class FastSync
       );
       if(newTarget.epoch > this.targetCheckpoint.epoch) {
         this.targetCheckpoint = newTarget;
-        this.setBlockImportTarget();
```

Comment:
This is to fix the `Received block parent root doesn't match our head` issue. This call would cause the next block range fetch while we're in the middle of current range.

Reply:
hmm but it won't trigger to sync forward if detects new finalized checkpoint (sync could stop now if new finalized epoch shows up while we are syncing current one)?
`Received block parent root doesn't match our head` - I don't think this happens because we are fetching next block range in middle of current one because thos functions are executed sequentially

Reply:
> hmm but it won't trigger to sync forward if detects new finalized checkpoint (sync could stop now if new finalized epoch shows up while we are syncing current one)?

if it detects new finalized checkpoint, it just set new `targetCheckpoin`, and `checkSyncProgress` will trigger the next sync

> I don't think this happens because we are fetching next block range in middle of current one because thos functions are executed sequentially

this is how it happens: I have `maxSlotImport` as 64 = 2 * SLOTS_PER_EPOCH. 
+ It fetch and process 142131...142162 `Fetching blocks for 142131...142162 slot range`
+ At slot 142144 (the middle of the range), chain processor detects new epoch and it emits `"processedCheckpoint"`
+ Then `checkSyncCompleted` called and it called `setBlockImportTarget` and it triggered a new block range as in the log `Fetching blocks for 142162...142225 slot range`

I think this is quite clear as we can see in the issue description #1183

Reply:
I guess, that's odd since since it iterates sequentially it cannot skip stuff, it can start fetching blocks for next batch but they ain't suppose to be processed before first batch.

in #1183 it says in logs we processed 142161 but call to head returns different one, so it apears we have some race condition between block post process and this  processSyncBlocks. We could call method for getting headRoot a bit later in processSyncBlocks method though (after we receive all blocks in range)

Reply:
> in #1183 it says in logs we processed 142161 but call to head returns different one

that's because we have `let headRoot = chain.forkChoice.headBlockRoot();` before `for await` so that `headRoot` became obsolete. The other way to fix the issue is to just have that call after `for await` to make sure we get up-to-date value, does that sound good to you?

Reply:
yeah, that was my suggestion too.

Still not sure about removing `this.setBlockImportTarget();` here. `checkSyncProgress` will most likely be called before `checkSyncCompleted` so it won't trigger new sync loop

Reply:
in this case, `checkSyncProgress` is not called before `checkSyncCompleted` because `checkSyncProgress` is called at the end of batch (which is 2 epochs) while `checkSyncProgress` is called in the middle (at end of 1st epoch)
+ we get batch 142131...142162
+ at 142144 `checkSyncCompleted` is called in the middle of the batch because it's new epoch
+ at 142162 `checkSyncProgress` is called (142131 + 2 * SLOTS_PER_EPOCH)

Reply:
if we configure `maxSlotImport = 1*SLOTS_PER_EPOCH` then yes, `checkSyncProgress` will most likely be called before `checkSyncCompleted`

Reply:
Yeah, but consider the case where we are one epoch from targeted finalized epoch. We would fetch missing blocks, call checkSyncProgress on last finalized block, then call checkSyncCompleted where new finalized target is detected from peers and sync would stall?

Reply:
> then call checkSyncCompleted where new finalized target is detected from peers 

if `newTarget epoch > targetCheckpoint.epoch` then we just set new targetCheckpoint, sync will continue for 1 more round
otherwise we stop the initial sync and initial sync is stopped and we transition to regular sync

Reply:
> Still not sure about removing this.setBlockImportTarget(); here. checkSyncProgress will most likely be called before checkSyncCompleted so it won't trigger new sync loop

from what I see, `checkSyncProgress` will always trigger new sync loop regardless of `targetCheckpoint`